server/src-lib/Hasura/RQL/GBoolExp.hs

```diff
@@ -109,16 +145,36 @@ parseOpExp parser fim (PGColInfo cn colTy _) (opStr, val) = case opStr of
     parseNsimilar = textOnlyOp colTy >> ANSIMILAR <$> parseOne
 
     parseIsNull   = bool ANISNOTNULL ANISNULL -- is null
-                    <$> decodeValue val
+                    <$> parseVal
+
     parseCeq      = CEQ <$> decodeAndValidateRhsCol
     parseCne      = CNE <$> decodeAndValidateRhsCol
     parseCgt      = CGT <$> decodeAndValidateRhsCol
     parseClt      = CLT <$> decodeAndValidateRhsCol
     parseCgte     = CGTE <$> decodeAndValidateRhsCol
     parseClte     = CLTE <$> decodeAndValidateRhsCol
 
+    jsonbOnlyOp m = case colTy of
+      PGJSONB -> m
+      ty      -> throwError $ buildMsg ty [PGJSONB]
+
+    parseGeometryOp f =
+      geometryOnlyOp colTy >> f <$> parseOne
+
+    onNothing mV errM = maybe errM return mV
+
+    parseSTDWithinObj = do
+      stdObj :: Object <- parseVal
+      distVal <- onNothing (M.lookup "distance" stdObj) $
```

Comment:
This looks like a job for aeson. Why not change 
```haskell
ASTDWithin a a
```
to 
```haskell
ASTDWithin (WithinOp a)
```
and `WithinOp` has both `distance` and `from` fields?